modules/core/src/commonMain/kotlin/fr/acinq/lightning/serialization/channel/Encryption.kt

```diff
@@ -95,9 +98,55 @@ object Encryption {
     /**
      * Convenience method that decrypts and deserializes a list of [PersistedChannelState] from an [EncryptedPeerStorage]
      */
-    fun PersistedChannelState.Companion.fromEncryptedPeerStorage(key: PrivateKey, encryptedPeerStorage: EncryptedPeerStorage): Result<Serialization.PeerStorageDeserializationResult> {
-        // we first assume that data is prefixed by 1 byte of serialization meta-info
+    fun PersistedChannelState.Companion.fromEncryptedPeerStorage(key: PrivateKey, encryptedPeerStorage: EncryptedPeerStorage, logger: MDCLogger?): Result<Serialization.PeerStorageDeserializationResult> {
+        logger?.info { "decrypting peer storage data using current method" }
+        val resCurrent = fromEncryptedPeerStorageCurrent(key, encryptedPeerStorage)
+        val resLegacy = if (resCurrent.isFailure) {
+            logger?.info { "decrypting peer storage data using legacy channel data method" }
+            fromEncryptedPeerStorageLegacyChannelData(key, encryptedPeerStorage)
+        } else null
+
+        return when {
+            resCurrent.isSuccess -> resCurrent
+            resLegacy != null && resLegacy.isSuccess -> resLegacy
+            else -> resCurrent
+        }
+    }
+
+    private fun fromEncryptedPeerStorageCurrent(key: PrivateKey, encryptedPeerStorage: EncryptedPeerStorage): Result<Serialization.PeerStorageDeserializationResult> {
+        // data is prefixed by 1 byte of serialization meta-info
         return runCatching { decrypt(key.value, encryptedPeerStorage.data.drop(1).toByteArray()) }
             .map { Serialization.deserializePeerStorage(encryptedPeerStorage.data[0], it) }
     }
+
+    /**
+     * Backward compatibility for an older backup mechanism where each channel data was stored individually, and restored from channel_reestablish messages.
+     */
+    private fun fromEncryptedPeerStorageLegacyChannelData(key: PrivateKey, encryptedPeerStorage: EncryptedPeerStorage): Result<Serialization.PeerStorageDeserializationResult> {
+        // The blob is a list of individually encrypted channel data (because each of them was previously stored separately)
+        val input = ByteArrayInput(encryptedPeerStorage.data.toByteArray())
+        val encryptedChannelDatas = input.readCollection { input.readDelimitedByteArray() }.toList()
+        val res: List<Result<Serialization.DeserializationResult>> = encryptedChannelDatas.map { data ->
+            runCatching { decrypt(key.value, data.drop(2).toByteArray()) }
+                .recoverCatching { decrypt(key.value, data) }
+                .map { Serialization.deserialize(it) }
```

Comment:
This is the same contant as the `fun PersistedChannelState.Companion.from()` function above, but that function will be deleted once the migration is done, so factoring wasn't worth it.